### PR TITLE
add command to repair broken filesystem trees

### PR DESCRIPTION
**Pull request by icewind1991**
If for any reason the path of an entry in the filecache doesn't match with it's expected path based on the path of it's parent node you end up with an entry in the filecache that exists in different places based on how the entry is generated. For example, while listing folder `/foo` is contains a file `bar.txt`, but when trying to do anything with `/foo/bar.txt` the file doesn't exists.

This command attempts to repair such entries by querying for entries there the path doesn't match the expected path based on it's parent path and filename and resets it's path to the expected one.

It is not guaranteed that the "new" path is the correct one, but while that would leave the filecache and actual filesystem out of sync, various automated and manual processes to fix that already exist as filecache synchronization is a much wider topic.
By comparison there isn't currently any way (short of manual database editing) to repair a broken file tree.

### CODE REVIEW

**Review by rullzer — APPROVED** [review]
Makes sense
But I didn't test it. Is there an easy way to test it?

**Comment by icewind1991** [comment]
you can manually break your filesystem tree by changing the `path` and `path_hash` of a filesystem entry without changing the parent

File: apps/files/lib/Command/RepairTree.php [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+		$output->writeln("Found " . count($rows) . " file entries with an invalid path");
+
+		if ($fix) {
+			$this->connection->beginTransaction();
+		}
+
+		$query = $this->connection->getQueryBuilder();
+		$query->update('filecache')
+			->set('path', $query->createParameter('path'))
+			->set('path_hash', $query->func()->md5($query->createParameter('path')))
```
Comment (nickvergessen) [inline apps/files/lib/Command/RepairTree.php:64]: I always wonder if this works on all DBs and basically the order of the "set" matter or its atomic or something.
Did you check that?

**Review by nickvergessen — APPROVED** [review]
Looks good

File: apps/files/lib/Command/RepairTree.php (hunk 1 above)
Comment (icewind1991) [reply apps/files/lib/Command/RepairTree.php:64]: the `path` here is not the column name, but the parameter name.

Comment (icewind1991) [reply apps/files/lib/Command/RepairTree.php:64]: When using a column name I think that at least some databases use the old value.

Comment (nickvergessen) [reply apps/files/lib/Command/RepairTree.php:64]: Right 👍🏼

**Review by ChristophWurst — APPROVED** [review]
Code makes sense as far as I understand the file cache

:+1: (:elephant:)

**Comment by PVince81** [comment]
would be good to have unit tests to confirm that this isn't causing any side effects

something like https://github.com/owncloud/core/pull/28253/files#diff-61c2bd81935832b24f87e329f7b6374aa736b3d7c74196581abe0892d983303e

**Pull request merged**